samples/compiler_plugins/xnnpack/src/xnnpack/Conversion/StablehloToXnnpack.cpp

```diff
@@ -14,20 +17,144 @@
+    auto outputType = info.output.getType().cast<RankedTensorType>();
+    if (!inputType.getElementType().isInteger(8) ||
+        !weightType.getElementType().isInteger(4) ||
+        !outputType.getElementType().isF32()) {
+      return error(
+          "unimplemented: fully connected layer without i8 input, i4 weight, "
+          "or f32 output");
+    }
+
+    if (info.needsTranspose) {
+      int64_t weightRank = weightType.getRank();
+      SmallVector<int64_t> perm(llvm::to_vector(llvm::seq(weightRank)));
+      std::swap(perm[perm.size() - 1], perm[perm.size() - 2]);
+      DenseIntElementsAttr permAttr = rewriter.getI64TensorAttr(perm);
+      info.weight = rewriter.create<stablehlo::TransposeOp>(
+          op.getLoc(), info.weight, permAttr);
+    }
+
+    Operation *outputDefiningOp = info.output.getDefiningOp();
+    if (outputDefiningOp != op) rewriter.eraseOp(op);
+    rewriter.replaceOpWithNewOp<Xnnpack::FullyConnectedNcQd8F32Qc4wOp>(
+        outputDefiningOp, outputType, info.input, info.weight);
     return success();
   }
+
+ private:
+  // Inputs and outputs of the matched fully connected layer pattern.
+  struct FullyConnectedInfo {
+    Value input;
+    Value weight;
+    Value output;
+    bool needsTranspose;  // Weight needs transpose before reduction `input_ij
+                          // weight_kj`
+  };
+
+  // Get the `stablehlo.dot_general`'s inputs before any casts and its output
+  // after any casts, and check if it matches the fully connected layer pattern.
+  static FailureOr<FullyConnectedInfo> getFullyConnectedInfo(
+      mlir::stablehlo::DotGeneralOp op,
+      llvm::function_ref<LogicalResult(std::string)> error) {
+    Value lhs = op.getLhs();
+    Value rhs = op.getRhs();
+    auto lhsType = lhs.getType().cast<RankedTensorType>();
+    auto rhsType = rhs.getType().cast<RankedTensorType>();
+
+    auto dotNumbers = op.getDotDimensionNumbers();
+    ArrayRef<int64_t> lhsContractingDims =
+        dotNumbers.getLhsContractingDimensions();
+    ArrayRef<int64_t> rhsContractingDims =
+        dotNumbers.getRhsContractingDimensions();
+    if (lhsContractingDims.size() != 1 || rhsContractingDims.size() != 1) {
+      return error("expected at most one contracting dimension");
+    }
+    int64_t lhsContractingDim = lhsContractingDims.front();
+    int64_t rhsContractingDim = rhsContractingDims.front();
+    int64_t lhsBatchDimsCount = lhsType.getRank() - 2;
+    int64_t rhsBatchDimsCount = rhsType.getRank() - 2;
+    if (lhsContractingDim - lhsBatchDimsCount != 1 ||
+        (rhsContractingDim - rhsBatchDimsCount != 0 &&
+         rhsContractingDim - rhsBatchDimsCount != 1)) {
+      return error("not a fully connected pattern");
+    }
+
+    FullyConnectedInfo info;
+    info.input = getUnconvertedValue(lhs);
+    info.weight = getUnconvertedValue(rhs);
+    info.output = getConvertedValue(op.getResult());
+    info.needsTranspose = rhsContractingDim - rhsBatchDimsCount == 0;
```

Comment:
Add a comment here or in the function comment about the expected shapes for inputs and when a transpose is needed.